src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,498 @@
+	op              baseOp
+	transformOpts   transform.Options
+	startBounds     block.Bounds
+	endBounds       block.Bounds
+	processedBlocks []bool
+}
+
+func newBlockCache(op baseOp, transformOpts transform.Options) *blockCache {
+	return &blockCache{
+		op:            op,
+		transformOpts: transformOpts,
+	}
+}
+func (c *blockCache) init(bounds block.Bounds) {
+	if c.initialized {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.initialized {
+		return
+	}
+
+	c.startBounds = bounds.Nearest(c.transformOpts.TimeSpec.Start)
+	c.endBounds = bounds.Nearest(c.transformOpts.TimeSpec.End.Add(-1 * bounds.StepSize))
+	numBlocks := c.endBounds.End().Sub(c.startBounds.Start) / bounds.Duration
+	c.blockList = make([]block.Block, numBlocks)
+	c.processedBlocks = make([]bool, numBlocks)
+	c.initialized = true
+}
+
+func (c *blockCache) index(t time.Time) (int, error) {
+	if t.Before(c.startBounds.Start) || t.After(c.endBounds.Start) {
+		return 0, fmt.Errorf("invalid time for the block cache: %v, start: %v, end: %v", t, c.startBounds.Start, c.endBounds.Start)
+	}
+
+	return int(t.Sub(c.startBounds.Start) / c.startBounds.Duration), nil
+}
+
+// Add the block to the cache, errors out if block already exists
+func (c *blockCache) add(key time.Time, b block.Block) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	index, err := c.index(key)
+	if err != nil {
+		return err
+	}
+
+	if c.blockList[index] != nil {
+		return fmt.Errorf("block already exists at index: %d", index)
+	}
+
+	c.blockList[index] = b
+	return nil
+}
+
+// Remove the block from the cache
+func (c *blockCache) remove(idx int) error {
+	if idx >= len(c.blockList) {
+		return fmt.Errorf("index out of range for remove: %d", idx)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.blockList[idx] = nil
+	return nil
+}
+
+// Get the block from the cache
+func (c *blockCache) get(key time.Time) (block.Block, bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	index, err := c.index(key)
+	if err != nil {
+		return nil, false
+	}
+
+	b := c.blockList[index]
+	return b, b != nil
+}
+
+// multiGet retrieves multiple blocks from the cache at once until if finds an empty block
+func (c *blockCache) multiGet(startBounds block.Bounds, numBlocks int, reverse bool) ([]block.Block, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	blks := make([]block.Block, 0, numBlocks)
+	if numBlocks == 0 {
+		return blks, nil
+	}
+
+	startIdx, err := c.index(startBounds.Start)
+	if err != nil {
+		return nil, err
+	}
+
+	// Process a single index
+	process := func(i int) (bool, error) {
+		if startIdx+i >= len(c.blockList) {
+			return true, fmt.Errorf("index out of range: %d", startIdx+i)
+		}
+
+		b := c.blockList[startIdx+i]
+		if b == nil {
+			return true, nil
+		}
+
+		blks = append(blks, b)
+		return false, nil
+	}
+
+	if reverse {
+		for i := numBlocks - 1; i >= 0; i-- {
+			empty, err := process(i)
+			if err != nil {
+				return nil, err
+			}
+
+			if empty {
+				break
+			}
+		}
+
+		reversed(blks)
```

Comment:
Can we use `sort.Reverse(blks)`?

Reply:
we don't want the blocks in descending sorted order, this is more about just reversing the array

Reply:
sort.Reverse just reverses the list, shouldn't do any sorting

Reply:
don't think that's correct. Check out : https://golang.org/pkg/sort/#Reverse. They have an example there. Reverse needs Len(), Comparator, etc. and sorts the list in reverse order

Reply:
Fair enough; weird that there's no inbuilt reverse array method